Add additional convenience methods to DecimalBigNumber

0xJem opened the pull request:
- Add `.eq` and `.lt`
- Add `toString` that takes options (decimals, formatting, trimming), and removes `toAccurateString` and `toFormattedString`
- Determine the number of decimal places from string input 
- More comprehensive tests 
- Don't require the decimal places to be specified in multiplication and division operations

0xJem committed:
Add support for initialising BigNumber with a number

0xJem committed:
Add copy function

0xJem committed:
Add .eq and .lt

0xJem committed:
Change toFormattedString to default to the DecimalBigNumber's configured decimal places, instead of 0

0xJem committed:
Correct tests

0xJem committed:
Add trimming of trailing zeroes

0xJem committed:
Test for trailing decimal point and zeroes in formatted strings

sam-potter commented:
> * Add support for initialising DecimalBigNumber with a number. BigNumber.js takes a number as input to the constructor, so this will smooth the transition to DecimalBigNumber

I don't really like this. Leaving out number as an input type was intentional when I first designed this. Basically, it's because we should only really be switching between BigNumbers (used for all internal calculations) and strings (used to display a number to the user, *and* importantly to also get input from the user). I'm kind of surprised that BigNumber encourages numbers as an input type anyway because if we use them we run into all the problems that BigNumber intends to avoid.
```js
 137123.31203981235.toString()
'137123.31203981236'
```
So I was hoping by only allowing strings, it would encourage developers to take their inputs from users as 
`<input type="text" />` rather than `<input type="number" />` to retain precision.

Also for the use case where you're comparing a regular number to a BigNumber. It makes more sense to fall "down" (in terms of precision) from DecimalBigNumber that it is to go "up", since we're likely doing something that is already imprecise.
```js
// Good
103123908 * new DecimalBigNumber("100000", 0).toApproxNumber()

// Worse (since it doesn't really gain anything in terms of precision)
new DecimalBigNumber(103123908, 0).mul(new DecimalBigNumber("100000", 0).toApproxNumber())
```

In terms of our transition from BigNumber, in the case where we are not passing in that BigNumber directly, worst case we just force a developer to do this:
```js
new DecimalBigNumber(103123908, 0)

// Should be done as
new DecimalBigNumber(103123908.toString(), 0)
```

Therefore meaning any imprecision that occurs is the developers fault, rather than DecimalBigNumbers fault from calling `toString` under the hood (which is a good thing because the latter would be a way more obscure bug).

> * Add function to copy

What's the use case for this?

> * Add `.eq` and `.lt`

Looks great, thanks for doing that.

> * `toFormattedString` defaults to the object instance's configured number of decimal places

I've reverted this change. When showing a number to the user, we pretty much only either want to show it perfectly accurately, or to a few decimals (probably max 4) with nice formatting (commas and stuff). So it feels better to default to 0 and have the developer "work up" towards a small number of decimal places, since that's the common use case. And because the current implementation of `toFormattedString` converts to a number first, if you format to 18 decimals, a number with a large enough precision would return something insanely inaccurate, so we shouldn't really do that by default.

> * Add `toFormattedStringTrimmed` to trim trailing zeroes

Great idea! Trimming the result will be really useful. Only thing is this function feels too similar to the original. So I've removed this in favour of adding an optional argument to `toFormattedString`, lmk what you think.

sam-potter committed:
Make `trimTrailingZeroes` an optional arg. Default formatting to 0

0xJem commented:
Thanks @sam-potter , the insight into your thought process was really useful.

On the constructor, I've added documentation so those assumptions/design decisions are clear.

copy: removed, as it's not needed for an immutable object.

toFormattedStringTrimmed: I was trying to figure out how to add a second, optional, named parameter (it's supported in Python, and I assumed it would be in JS/TS), but this is a good workaround.

toFormattedString: my issue is that this forces imprecision (dropping the decimal places), when our goal is to maintain it. It feels more natural for the number shown to be whatever number of decimal places was specified at initialisation, instead of `toFormattedString` making the decision for me.

0xJem committed:
Document design decisions, add tests

0xJem committed:
Add test cases

0xJem commented:
> Also for the use case where you're comparing a regular number to a BigNumber. It makes more sense to fall "down" (in terms of precision) from DecimalBigNumber that it is to go "up", since we're likely doing something that is already imprecise.
> 
> ```js
> // Good
> 103123908 * new DecimalBigNumber("100000", 0).toApproxNumber()
> 
> // Worse (since it doesn't really gain anything in terms of precision)
> new DecimalBigNumber(103123908, 0).mul(new DecimalBigNumber("100000", 0).toApproxNumber())
> ```

Isn't the objective to retain precision as much as possible? I don't see how it's advisable to drop down to `number` to perform the multiplication. Shouldn't we be recommending to store the number (e.g. 103123908) as a DecimalBigNumber in the first place, use that in the multiplication, and keep the result as a DecimalBigNumber?

0xJem commented:
I've pushed a few documentation and test changes for your review.

sam-potter commented:
> toFormattedString: my issue is that this forces imprecision (dropping the decimal places), when our goal is to maintain it. It feels more natural for the number shown to be whatever number of decimal places was specified at initialisation, instead of `toFormattedString` making the decision for me.

I feel like the ideal solution, is probably to incorporate formatting (i.e. adding commas, and choosing a specific number of decimal places to be shown to the user) into the `toAccurateString` method. If we could do that without losing accuracy (i.e. we don't convert to a number at all whilst formatting) then it would make sense to default to whatever decimal places was specified at initialisation. That would also probably allow us to remove the `toFormattedString` method entirely, and perhaps even rename the `toAccurateString` method to just `toString`, since its accuracy would be guaranteed regardless if the number is formatted or not. What are your thoughts on this? 

> Isn't the objective to retain precision as much as possible? I don't see how it's advisable to drop down to `number` to perform the multiplication. Shouldn't we be recommending to store the number (e.g. 103123908) as a DecimalBigNumber in the first place, use that in the multiplication, and keep the result as a DecimalBigNumber?

I've changed my mind, I think you're absolutely right. When I wrote that example, I was thinking of the case where that `number` was something like a price returned from coingecko, so it could be initially imprecise. But even if that `number` *is* initially imprecise, we definitely should be encouraging converting and storing it as a DecimalBigNumber, because at least we can be certain that any calculations therein after won't worsen that possible initial imprecision.

One idea i've just thought of that might be helpful in the case of turning strings into a DecimalBigNumber, is to make specifying the decimal value optional, and instead just infer the decimal value as how ever many decimals are provided on initialisation. I.e.

```js
new DecimalBigNumber('100.312') // Could be inferred as a 3 decimal number
```

Because a number like a price doesn't have a specific set of decimal places that it should be (unlike token decimals do), so it would be nice to infer the number of decimal places as the maximum of whatever string is provided. What are your thoughts on this also?

sam-potter commented:
Also, the documentation and comments are awesome. Thanks for doing that

0xJem commented:
> Because a number like a price doesn't have a specific set of decimal places that it should be (unlike token decimals do), so it would be nice to infer the number of decimal places as the maximum of whatever string is provided. What are your thoughts on this also?

Makes sense.

> I feel like the ideal solution, is probably to incorporate formatting (i.e. adding commas, and choosing a specific number of decimal places to be shown to the user) into the toAccurateString method. If we could do that without losing accuracy (i.e. we don't convert to a number at all whilst formatting) then it would make sense to default to whatever decimal places was specified at initialisation. That would also probably allow us to remove the toFormattedString method entirely, and perhaps even rename the toAccurateString method to just toString, since its accuracy would be guaranteed regardless if the number is formatted or not. What are your thoughts on this?

Something like this?
`public toString((options?: {
                               decimals?: number,
                               trim?: boolean, // true
                               format?: boolean = false,
}): string`

0xJem committed:
Make the decimals parameter optional and determine the number of decimal places

0xJem committed:
WIP implementation of consolidated toString() function

0xJem committed:
Implemented omitting and padding of decimals

0xJem committed:
Add number formatting. ALL GREEN.

0xJem committed:
Add test cases for toApproxNumber()

0xJem committed:
Replace DecimalBigNumber usage of toAccurateString()

0xJem committed:
Remove toAccurateString()

0xJem committed:
Remove toFormattedString()

0xJem committed:
Merge branch 'develop' into bignumber-number-input

# Conflicts:
#	src/views/Stake/components/StakeArea/components/StakeInputArea/StakeInputArea.tsx

0xJem committed:
Update test snapshots

0xJem committed:
Update test snapshots

0xJem commented:
@sam-potter Your feedback has been integrated. Please review!

0xJem committed:
Fix undefined error

sam-potter committed:
Add constructor overloads

sam-potter committed:
Parse input before string inference

sam-potter committed:
Make args default to `{}`

sam-potter committed:
Prefer ethers `commify` method

sam-potter committed:
Refactor decimal parsing implementation

sam-potter committed:
Slight implementation changes

sam-potter committed:
Make decimal values optional

sam-potter committed:
Default to 0 rather than throw

sam-potter committed:
Added `_ensurePositive` method

sam-potter committed:
Refactor

sam-potter commented:
Looks awesome thanks for doing all that. I just made a couple changes:

- TS now enforces that if you pass in a BigNumber you must pass when a decimal value when initialising, but remains optional when using a string, with function overloads.
- Prefer use of ethers `commify` method to add thousands separators
- Added the `@deprecated` tag to `toApproxNumber` since we shouldn't really encourage its use anymore.
- Made the decimal param in the `mul` and `div` functions optional, and default to the smallest of the two decimal values when performing those operations. As I've been using this in the react-query bond page update, I realised having this as a required param was annoying/didn't seem useful.
- If a negative decimal value is provided anywhere, it'll default to 0 rather than throw an error.

Lmk your thoughts on these

0xJem committed:
mul/div test cases without decimals, and fix an edge case

0xJem commented:
> If a negative decimal value is provided anywhere, it'll default to 0 rather than throw an error.

Couldn't that lead to unexpected results?

0xJem committed:
Maintain precision when multiplying

0xJem committed:
Add more comprehensive tests on division, and a note about precision

sam-potter committed:
Typo

sam-potter committed:
Implementation changes

sam-potter committed:
Consolidate terminology

sam-potter committed:
cs

sam-potter commented:
I've removed the second argument to `mul(value, precision)` (since it now returns the maximum precision of the result) in favour of just calling toString({ precision: ... }) on the result.

Also, I've tried to consolidate the terminology by using "value" instead of "number" and "precision" instead of "decimals", lmk if you think I've made that more confusing.

If you think that all looks good, I'll go and fix all the other parts of the codebase I would've broken with these changes

sam-potter committed:
Rename "precision" back to "decimals"

sam-potter committed:
Update API use in codebase

0xJem commented:
@appleseed-iii for your kind review, ser

appleseed-iii commented on src/helpers/DecimalBigNumber/DecimalBigNumber.ts:
```diff
@@ -1,143 +1,260 @@
 
   /**
-   * Determines if this number is greater than the provided value
+   * Determines if this va;ie is greater than the provided value
```
is this supposed to say `value`

0xJem replied:
Yes it should say value

appleseed-iii reviewed (approved):
the documentation in this library you both built is fantastic 🎸 🎵 .
I had one minor comment in here to verify.

0xJem committed:
Fix typo
